File: Frontend/src/sdk/pedidos.js
import { Pedido } from './models'

export class PedidosClient {
  constructor(http, rootClient) {
    this._http = http
    this._root = rootClient
  }

  async add(payload) {
    return this.create(payload)
  }

  async create(payload) {
    const data = await this._http.post('/pedidos/', payload)
    return new Pedido(this._root, data)
  }

  async mine() {
    const items = await this._http.get('/pedidos/mis')
    return items.map((p) => new Pedido(this._root, p))
  }

  async list(options = {}) {
    const params = new URLSearchParams()
    if (options.skip != null) params.set('skip', String(options.skip))
    if (options.limit != null) params.set('limit', String(options.limit))
    if (options.estado != null && String(options.estado).trim()) {
      params.set('estado', String(options.estado).trim().toLowerCase())
    }
    const query = params.toString()
    const url = query ? `/pedidos/?${query}` : '/pedidos/'
    const items = await this._http.get(url)
    return items.map((p) => new Pedido(this._root, p))
  }

  async get(id) {
    const data = await this._http.get(`/pedidos/${id}`)
    return new Pedido(this._root, data)
  }

  async updateEstado(id, payload) {
    const data = await this._http.patch(`/pedidos/${id}/estado`, payload)
    return new Pedido(this._root, data)
  }

  async metrics(options = {}) {
    const params = new URLSearchParams()

    if (options.rango != null) params.set('rango', String(options.rango))
    if (options.fecha_desde != null) params.set('fecha_desde', String(options.fecha_desde))
    if (options.fecha_hasta != null) params.set('fecha_hasta', String(options.fecha_hasta))
    if (options.group_by != null) params.set('group_by', String(options.group_by))

    if (options.estado != null) {
      const estado = Array.isArray(options.estado)
        ? options.estado.join(',')
        : String(options.estado)
      if (estado.trim()) params.set('estado', estado)
    }

    if (options.pais != null) {
      const pais = Array.isArray(options.pais) ? options.pais.join(',') : String(options.pais)
      if (pais.trim()) params.set('pais', pais)
    }

    if (options.producto_ids != null) {
      const productoIds = Array.isArray(options.producto_ids)
        ? options.producto_ids.join(',')
        : String(options.producto_ids)
      if (productoIds.trim()) params.set('producto_ids', productoIds)
    }

    if (options.categoria_ids != null) {
      const categoriaIds = Array.isArray(options.categoria_ids)
        ? options.categoria_ids.join(',')
        : String(options.categoria_ids)
      if (categoriaIds.trim()) params.set('categoria_ids', categoriaIds)
    }

    if (options.min_total != null && String(options.min_total).trim() !== '') {
      params.set('min_total', String(options.min_total))
    }
    if (options.max_total != null && String(options.max_total).trim() !== '') {
      params.set('max_total', String(options.max_total))
    }
    if (options.top_n != null) params.set('top_n', String(options.top_n))
    if (options.ventas_solo_aprobadas != null) {
      params.set('ventas_solo_aprobadas', String(Boolean(options.ventas_solo_aprobadas)))
    }

    const query = params.toString()
    const url = query ? `/pedidos/metricas?${query}` : '/pedidos/metricas'
    return this._http.get(url)
  }

  async delete(id) {
    return this._http.delete(`/pedidos/${id}`)
  }
}

File: Backend/app/api/routes/pedidos.py
from collections.abc import Sequence
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.dependencies.db import get_db
from app.api.dependencies.security import (
	get_current_admin,
	get_current_user,
	get_current_user_optional,
)
from app.controllers import pedidos as pedidos_controller
from app.schemas.metrics import PedidosMetricasResponse, RegistroVisitaPayload, RegistroVisitaResponse
from app.schemas.tickets import PedidoCreate, PedidoEstadoUpdate, PedidoResponse
from app.services.visitas_metricas import registrar_visita


router = APIRouter(prefix="/pedidos", tags=["pedidos"])


@router.post("/", response_model=PedidoResponse, status_code=201)
async def crear_pedido(
	pedido_in: PedidoCreate,
	current_user=Depends(get_current_user_optional),
	db: Session = Depends(get_db),
):
	try:
		return pedidos_controller.crear_pedido_para_usuario(
			db,
			usuario_id=current_user.id if current_user is not None else None,
			pedido_in=pedido_in,
		)
	except ValueError as exc:
		raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/mis", response_model=Sequence[PedidoResponse])
async def listar_mis_pedidos(
	current_user=Depends(get_current_user),
	db: Session = Depends(get_db),
):
	return pedidos_controller.listar_pedidos_de_usuario(db, usuario_id=current_user.id)


@router.get("/", response_model=Sequence[PedidoResponse])
async def listar_pedidos_admin(
	skip: int = 0,
	limit: int = 50,
	estado: str | None = None,
	db: Session = Depends(get_db),
	admin=Depends(get_current_admin),
):
	_ = admin
	return pedidos_controller.listar_pedidos(db, skip=skip, limit=limit, estado=estado)


@router.get("/metricas", response_model=PedidosMetricasResponse)
async def obtener_metricas_pedidos_admin(
	rango: str = Query(default="30d"),
	fecha_desde: datetime | None = Query(default=None),
	fecha_hasta: datetime | None = Query(default=None),
	group_by: str = Query(default="day"),
	estado: str | None = Query(default=None),
	pais: str | None = Query(default=None),
	producto_ids: str | None = Query(default=None),
	categoria_ids: str | None = Query(default=None),
	min_total: float | None = Query(default=None, ge=0),
	max_total: float | None = Query(default=None, ge=0),
	top_n: int = Query(default=8, ge=3, le=20),
	ventas_solo_aprobadas: bool = Query(default=True),
	db: Session = Depends(get_db),
	admin=Depends(get_current_admin),
):
	_ = admin
	return pedidos_controller.obtener_metricas_pedidos(
		db,
		rango=rango,
		fecha_desde=fecha_desde,
		fecha_hasta=fecha_hasta,
		group_by=group_by,
		estado=estado,
		pais=pais,
		producto_ids=producto_ids,
		categoria_ids=categoria_ids,
		min_total=min_total,
		max_total=max_total,
		top_n=top_n,
		ventas_solo_aprobadas=ventas_solo_aprobadas,
	)


@router.post("/metricas/visitas/register", response_model=RegistroVisitaResponse, status_code=status.HTTP_202_ACCEPTED)
async def registrar_visita_publica(payload: RegistroVisitaPayload):
	accepted = registrar_visita(visitor_id=payload.visitor_id, path=payload.path)
	return RegistroVisitaResponse(accepted=accepted)


@router.get("/{pedido_id}", response_model=PedidoResponse)
async def get_pedido(
	pedido_id: int,
	current_user=Depends(get_current_user),
	db: Session = Depends(get_db),
):
	pedido = pedidos_controller.get_pedido(db, pedido_id=pedido_id)
	if pedido is None:
		raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pedido no encontrado")

	if not current_user.is_admin and pedido.usuario_id != current_user.id:
		raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="No autorizado")

	return pedido


@router.patch("/{pedido_id}/estado", response_model=PedidoResponse)
async def actualizar_estado_pedido(
	pedido_id: int,
	payload: PedidoEstadoUpdate,
	db: Session = Depends(get_db),
	admin=Depends(get_current_admin),
):
	_ = admin
	try:
		pedido = pedidos_controller.actualizar_estado_pedido(
			db,
			pedido_id=pedido_id,
			nuevo_estado=payload.estado,
			usuario_id_asignacion=payload.usuario_id,
		)
	except ValueError as exc:
		raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

	if pedido is None:
		raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pedido no encontrado")

	return pedido


@router.delete("/{pedido_id}", status_code=204)
async def eliminar_pedido(
	pedido_id: int,
	db: Session = Depends(get_db),
	admin=Depends(get_current_admin),
):
	_ = admin
	deleted = pedidos_controller.eliminar_pedido(db, pedido_id=pedido_id)
	if not deleted:
		raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pedido no encontrado")
	return None
